***Simple Neural Network with Keras Sequential API***

In [4]:
!pip install tensorflow


  Using cached astunparse-1.6.3-py2.py3-none-any.whl.metadata (4.4 kB)
  Using cached flatbuffers-25.2.10-py2.py3-none-any.whl.metadata (875 bytes)
  Using cached google_pasta-0.2.0-py3-none-any.whl.metadata (814 bytes)
  Using cached libclang-18.1.1-1-py2.py3-none-macosx_11_0_arm64.whl.metadata (5.2 kB)
  Using cached opt_einsum-3.4.0-py3-none-any.whl.metadata (6.3 kB)
  Using cached tensorboard_data_server-0.7.2-py3-none-any.whl.metadata (1.1 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 252.7/252.7 MB 19.8 MB/s eta 0:00:0000:0100:01
Using cached astunparse-1.6.3-py2.py3-none-any.whl (12 kB)
Using cached flatbuffers-25.2.10-py2.py3-none-any.whl (30 kB)
Using cached google_pasta-0.2.0-py3-none-any.whl (57 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 10.6/10.6 MB 21.9 MB/s eta 0:00:00 0:00:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.4/1.4 MB 15.2 MB/s eta 0:00:00
Using cached libclang-18.1.1-1-py2.py3-none-macosx_11_0_arm64.whl (25.8 MB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

In [5]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
import numpy as np

# Generate some random data
x_train = np.random.random((1000, 20))  # 1000 samples, 20 features each
y_train = np.random.randint(2, size=(1000, 1))  # Binary labels (0 or 1)

x_test = np.random.random((200, 20))  # 200 test samples
y_test = np.random.randint(2, size=(200, 1))  # Binary labels for testing

# Build a Sequential model
model = Sequential()

# Add a hidden layer with 64 neurons and ReLU activation
model.add(Dense(64, activation='relu', input_shape=(20,)))

# Add another hidden layer with 32 neurons and ReLU activation
model.add(Dense(32, activation='relu'))

# Add an output layer with 1 neuron and sigmoid activation for binary classification
model.add(Dense(1, activation='sigmoid'))

# Compile the model
model.compile(optimizer='adam',
              loss='binary_crossentropy',
              metrics=['accuracy'])

# Train the model
model.fit(x_train, y_train, epochs=5, batch_size=32)

# Evaluate the model on the test data
test_loss, test_acc = model.evaluate(x_test, y_test)
print(f'Test accuracy: {test_acc}')


Epoch 1/5


/opt/anaconda3/lib/python3.12/site-packages/keras/src/layers/core/dense.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


32/32 ━━━━━━━━━━━━━━━━━━━━ 0s 630us/step - accuracy: 0.5168 - loss: 0.6989
Epoch 2/5
32/32 ━━━━━━━━━━━━━━━━━━━━ 0s 621us/step - accuracy: 0.5414 - loss: 0.6894
Epoch 3/5
32/32 ━━━━━━━━━━━━━━━━━━━━ 0s 586us/step - accuracy: 0.5728 - loss: 0.6874
Epoch 4/5
32/32 ━━━━━━━━━━━━━━━━━━━━ 0s 551us/step - accuracy: 0.5513 - loss: 0.6846
Epoch 5/5
32/32 ━━━━━━━━━━━━━━━━━━━━ 0s 499us/step - accuracy: 0.5983 - loss: 0.6797
7/7 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.5138 - loss: 0.6917 
Test accuracy: 0.5199999809265137


***Sequential API to create and train a neural network for classifying the MNIST dataset.***

In [6]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Flatten
from tensorflow.keras.datasets import mnist
from tensorflow.keras.utils import to_categorical

# Load the MNIST dataset
(x_train, y_train), (x_test, y_test) = mnist.load_data()

# Preprocess the data: normalize images and one-hot encode labels
x_train = x_train.astype('float32') / 255.0
x_test = x_test.astype('float32') / 255.0

y_train = to_categorical(y_train, 10)
y_test = to_categorical(y_test, 10)

# Build a Sequential model
model = Sequential()

# Flatten the input (28x28 images) into a vector of size 784
model.add(Flatten(input_shape=(28, 28)))

# Add a hidden layer with 128 neurons and ReLU activation
model.add(Dense(128, activation='relu'))

# Add the output layer with 10 neurons (one for each class) and softmax activation
model.add(Dense(10, activation='softmax'))

# Compile the model
model.compile(optimizer='adam',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train the model
model.fit(x_train, y_train, epochs=5, batch_size=32, validation_split=0.2)

# Evaluate the model on the test data
test_loss, test_acc = model.evaluate(x_test, y_test)
print(f'Test accuracy: {test_acc}')


11490434/11490434 ━━━━━━━━━━━━━━━━━━━━ 1s 0us/step
Epoch 1/5


/opt/anaconda3/lib/python3.12/site-packages/keras/src/layers/reshaping/flatten.py:37: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


1500/1500 ━━━━━━━━━━━━━━━━━━━━ 1s 842us/step - accuracy: 0.8682 - loss: 0.4602 - val_accuracy: 0.9566 - val_loss: 0.1514
Epoch 2/5
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 1s 721us/step - accuracy: 0.9620 - loss: 0.1328 - val_accuracy: 0.9668 - val_loss: 0.1174
Epoch 3/5
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 1s 743us/step - accuracy: 0.9743 - loss: 0.0872 - val_accuracy: 0.9669 - val_loss: 0.1132
Epoch 4/5
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 1s 766us/step - accuracy: 0.9820 - loss: 0.0627 - val_accuracy: 0.9726 - val_loss: 0.0951
Epoch 5/5
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 1s 804us/step - accuracy: 0.9866 - loss: 0.0464 - val_accuracy: 0.9736 - val_loss: 0.0922
313/313 ━━━━━━━━━━━━━━━━━━━━ 0s 346us/step - accuracy: 0.9738 - loss: 0.0848
Test accuracy: 0.9769999980926514


# ICP6 – Programming Practice: Build Your Own Neural Network

### Objective:
Build and evaluate a feedforward neural network using Keras Sequential API.

### Task Requirements:
1. Build a Sequential model with at least **5 hidden layers** using **ReLU** activation.
2. Use **Softmax** activation in the output layer for multi-class classification.
3. Experiment with:
   - More layers / different neuron counts
   - Different activation functions (`tanh`, `sigmoid`)
   - Different optimizers (`adam`, `sgd`, `rmsprop`)
4. Use the **MNIST dataset**.
5. Achieve **test accuracy > 99%**.


In [7]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Flatten, Input
from tensorflow.keras.datasets import mnist
from tensorflow.keras.utils import to_categorical
import pandas as pd


What’s Happening:
You import the necessary packages:

tensorflow and keras: for building and training neural networks

mnist: a dataset of 28x28 handwritten digits (0–9)

to_categorical: converts labels like 3 into [0, 0, 0, 1, 0, ..., 0]

pandas: helps organize results in tables

In [8]:
(x_train, y_train), (x_test, y_test) = mnist.load_data()
x_train = x_train.astype('float32') / 255.0
x_test = x_test.astype('float32') / 255.0
y_train = to_categorical(y_train, 10)
y_test = to_categorical(y_test, 10)


In [ ]:
What’s Happening:
You load the MNIST digit dataset

Normalize pixel values from 0–255 to 0–1 for faster training

Convert numeric labels (0–9) into one-hot encoded vectors (needed for softmax)



In [9]:
def build_model(activation='relu', optimizer='adam'):
    model = Sequential()
    model.add(Input(shape=(28, 28)))
    model.add(Flatten())
    model.add(Dense(256, activation=activation))
    model.add(Dense(128, activation=activation))
    model.add(Dense(64, activation=activation))
    model.add(Dense(32, activation=activation))
    model.add(Dense(16, activation=activation))
    model.add(Dense(10, activation='softmax'))
    model.compile(optimizer=optimizer, loss='categorical_crossentropy', metrics=['accuracy'])
    return model


In [ ]:
You're defining a function to build a model with 5 hidden layers

All hidden layers use activation='relu' (default)

The final layer uses softmax to predict probability for 10 digit classes

You can also change the activation function or optimizer when calling the function

In [10]:
model = build_model()
model.fit(x_train, y_train, epochs=5, batch_size=64, validation_split=0.2)
test_loss, test_acc = model.evaluate(x_test, y_test)
print(f"Test accuracy: {test_acc:.4f}")


Epoch 1/5
750/750 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8060 - loss: 0.6314 - val_accuracy: 0.9560 - val_loss: 0.1443
Epoch 2/5
750/750 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.9648 - loss: 0.1178 - val_accuracy: 0.9686 - val_loss: 0.1035
Epoch 3/5
750/750 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.9772 - loss: 0.0747 - val_accuracy: 0.9691 - val_loss: 0.1056
Epoch 4/5
750/750 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.9820 - loss: 0.0592 - val_accuracy: 0.9709 - val_loss: 0.1037
Epoch 5/5
750/750 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.9867 - loss: 0.0426 - val_accuracy: 0.9749 - val_loss: 0.0920
313/313 ━━━━━━━━━━━━━━━━━━━━ 0s 411us/step - accuracy: 0.9741 - loss: 0.1012
Test accuracy: 0.9765


In [ ]:
You train the model for 5 cycles (epochs) on 80% of training data

It uses 20% for validation to see how well the model is generalizing

Then you evaluate it on the unseen test data

Finally, it prints the accuracy — you aim for >99%

In [11]:
results = []
activations = ['relu', 'tanh', 'sigmoid']
optimizers = ['adam', 'sgd', 'rmsprop']

for act in activations:
    for opt in optimizers:
        print(f"Training with activation={act}, optimizer={opt}")
        model = build_model(activation=act, optimizer=opt)
        model.fit(x_train, y_train, epochs=5, batch_size=64, validation_split=0.2, verbose=0)
        _, test_acc = model.evaluate(x_test, y_test, verbose=0)
        results.append({
            'Activation': act,
            'Optimizer': opt,
            'Test Accuracy (%)': round(test_acc * 100, 2)
        })


Training with activation=relu, optimizer=adam
Training with activation=relu, optimizer=sgd
Training with activation=relu, optimizer=rmsprop
Training with activation=tanh, optimizer=adam
Training with activation=tanh, optimizer=sgd
Training with activation=tanh, optimizer=rmsprop
Training with activation=sigmoid, optimizer=adam
Training with activation=sigmoid, optimizer=sgd
Training with activation=sigmoid, optimizer=rmsprop


In [ ]:
You test all 9 combinations of activations and optimizers

For each combination, it trains the model, evaluates accuracy, and stores the results

In [12]:
df = pd.DataFrame(results)
df.sort_values(by='Test Accuracy (%)', ascending=False)


,Activation,Optimizer,Test Accuracy (%)
2,relu,rmsprop,97.74
0,relu,adam,97.05
5,tanh,rmsprop,96.92
3,tanh,adam,96.39
6,sigmoid,adam,95.35
1,relu,sgd,94.31
4,tanh,sgd,93.52
8,sigmoid,rmsprop,90.16
7,sigmoid,sgd,11.35


## ✅ Conclusion:

- Best performance achieved with `ReLU` + rmsprop`, reaching over 97% accuracy.
- `Sigmoid` underperformed due to vanishing gradient.
- Experiments highlight how activation functions and optimizers impact learning speed and accuracy.
